src/java.desktop/share/classes/com/sun/imageio/plugins/jpeg/ExifMarkerSegment.java

```diff
@@ -0,0 +1,294 @@
+
+    private static final int TIFF_BIG_ENDIAN = 0x4d4d;
+    private static final int TIFF_MAGIC = 42;
+    private static final int TIFF_TYPE_SHORT = 3;
+    private static final int TAG_IMAGE_WIDTH = 256;
+    private static final int TAG_IMAGE_HEIGHT = 257;
+    private static final int TAG_DATE_TIME = 306;
+    private static final int TAG_JPEG_INTERCHANGE_FORMAT = 513;
+    private static final int TAG_JPEG_INTERCHANGE_FORMAT_LENGTH = 514;
+
+    int thumbnailPos = -1;
+    int thumbnailLength = -1;
+    boolean isThumbnailJPEG;
+
+    int thumbnailWidth = -1;
+    int thumbnailHeight = -1;
+
+    final long firstIFDOffset;
+    final List<ImageFileDirectory> imageFileDirectories = new LinkedList<>();
+
+    ExifMarkerSegment(MarkerSegment originalSegment) throws IOException {
+        super(originalSegment.tag);
+        this.length = originalSegment.length;
+        this.data = originalSegment.data;
+
+        ByteArrayInputStream in = new ByteArrayInputStream(data, 6, data.length - 6);
+
+        // we aren't actually going to read anything as an image yet, but ImageInputStream
+        // has useful helper methods:
+        ImageInputStream input = new MemoryCacheImageInputStream(in);
+        input.setByteOrder(input.readUnsignedShort() == TIFF_BIG_ENDIAN ?
+                ByteOrder.BIG_ENDIAN : ByteOrder.LITTLE_ENDIAN);
+        if (input.readUnsignedShort() != TIFF_MAGIC) {
+            throw new IllegalArgumentException("Bad magic number");
+        }
+
+        firstIFDOffset = input.readUnsignedInt();
+        ImageFileDirectory ifd1 = null;
+        ImageFileDirectory ifd2 = null;
+        if (firstIFDOffset != 0) {
+            ifd1 = new ImageFileDirectory(input, firstIFDOffset);
+            imageFileDirectories.add(ifd1);
+
+            long secondIFDOffset = ifd1.nextIFD;
+            if (secondIFDOffset != 0) {
+                ifd2 = new ImageFileDirectory(input, secondIFDOffset);
+                imageFileDirectories.add(ifd2);
+            }
+        }
+
+        if (ifd2 != null) {
+            // the thumbnail should always be described in the 2nd IFD (if it exists at all)
+
+            thumbnailPos = ifd2.getTagValueAsInt(TAG_JPEG_INTERCHANGE_FORMAT);
+            thumbnailLength = ifd2.getTagValueAsInt(TAG_JPEG_INTERCHANGE_FORMAT_LENGTH);
+            if (thumbnailPos != NO_VALUE && thumbnailLength != NO_VALUE) {
+                // The `compression` tag (259) should also help inform whether we read this
+                // image as a JPEG or TIFF. But in reality this is tricky: the docs say
+                // the value for a JPEG encoding is 0x0006, but the `jdk_8160327-plastic-wrap.jpg`
+                // file shows it can also sometimes be 0x60000. I've also observed it to be
+                // undefined, 0x0007, or several variations of 0x????0006. Similarly the same
+                // tag should be 0x0001 for TIFFs, but I also observed a case where it as 0x10000.
```

Comment:
Isn't this ( 0x0001/ 0x0006vs 0x1000/0x6000) just a matter of endianness in the TIFF structure? Some odd writers may also use LONG/32 bit values, even though the TiFF and Exif specs only mention SHORT/16 bit values for the compression tag. 

Compression 7 "New JPEG"  is not as per the Exif spec, but it can probably safely be treated the same way as "Old JPEG" compression 6 for Exif thumbnails.

Reply:
Yes, it probably is endianness, or endianness-related. My first design question is: should we care? Currently this PR infers whether we're looking for a JPEG or TIFF thumbnail based on other fields. If we strictly rely on the compression tag (250) instead: is that better/desirable? (That is: we could just throw an IOException in the rare case this field is missing/broken.)

Reply:
> Compression 7 "New JPEG" is not as per the Exif spec, but it can probably safely be treated the same way as "Old JPEG" compression 6 for Exif thumbnails.

Probably the `Compression` tag should not be relied upon. While the Exif specification strangely states that for compressed thumbnails its value should be 6, there is no harm in its being 7.

Reply:
> Currently this PR infers whether we're looking for a JPEG or TIFF thumbnail based on other fields.

Specifically the `JPEGInterchangeFormat` and `JPEGInterchangeFormatLength` fields. In the Exif 3.0 specification these have support level _mandatory_ for compressed thumbnails and _disallowed_ for uncompressed thumbnails.

> If we strictly rely on the compression tag (250) instead: is that better/desirable?

I don't think so.

> (That is: we could just throw an IOException in the rare case this field is missing/broken.)

This will indeed happen if `JPEGInterchangeFormat` does not point to a valid JPEG stream (SOI marker).